File: tests/r_parity/08_dml.py
"""StatsPAI DML PLR parity (Python side) -- Module 08.

Runs sp.dml(model='plr') with linear-regression nuisance learners
on the Card 1995 replica. Cross-fitting uses 5 folds with a fixed
seed. The companion 08_dml.R uses DoubleML::DoubleMLPLR with
mlr3::regr.lm and an identical fold split.

Tolerance: rel < 1e-3. With deterministic linear nuisance learners
the only Monte Carlo source is the fold split itself, so the gap
should be small.
"""
from __future__ import annotations

import numpy as np
import statspai as sp
from sklearn.linear_model import LinearRegression

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "08_dml"
COVARIATES = ["exper", "expersq", "black", "south", "smsa"]


def main() -> None:
    # simulated=True is load-bearing, not decoration. This module *writes*
    # data/08_dml.csv, and the committed fixture -- together with the R and
    # Stata goldens computed on it -- is the calibrated replica. The
    # loader's default later became the real extract, so a bare
    # card_1995() call silently regenerates the fixture from different
    # data and breaks the same-byte premise of the row. Original-data
    # Card parity is a separate ledger (tests/orig_parity/01_card_original).
    df = sp.datasets.card_1995(simulated=True)
    df["fold_id"] = np.arange(len(df)) % 5
    dump_csv(df, MODULE)

    # NOTE: pass numpy seed so sklearn's KFold split is deterministic;
    # sp.dml accepts learners through model_y / model_d kwargs.
    np.random.seed(PARITY_SEED)
    fit = sp.dml(
        data=df, y="lwage", d="educ", X=COVARIATES,
        model="plr",
        model_y=LinearRegression(),
        model_d=LinearRegression(),
        n_folds=5,
        fold_indices=df["fold_id"].to_numpy(),
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE, side="py", statistic="theta_DML_PLR",
            estimate=float(fit.estimate),
            se=float(fit.se),
            ci_lo=float(fit.ci[0]) if fit.ci is not None else None,
            ci_hi=float(fit.ci[1]) if fit.ci is not None else None,
            n=int(fit.n_obs),
        )
    ]

    write_results(
        MODULE, "py", rows,
        extra={
            "dml_model": "PLR",
            "n_folds": int(fit.model_info["n_folds"]),
            "ml_g": "LinearRegression",
            "ml_m": "LinearRegression",
            "seed": PARITY_SEED,
            "fold_source": str(fit.model_info["fold_source"]),
            "fold_column": "fold_id",
            "fold_parity_note": (
                "Python and R read the same fold_id column and use "
                "explicit sample-splitting APIs before fitting the "
                "linear-nuisance PLR score."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/08_dml.R
# StatsPAI DML PLR parity (R side) -- Module 08.
#
# Reads data/08_dml.csv (the StatsPAI Card 1995 replica) and runs
# DoubleML::DoubleMLPLR with mlr3::regr.lm nuisance learners. Five
# folds, single repetition. Tolerance: rel < 1e-3.
#
# The CSV carries a deterministic fold_id column written by the Python
# side.  Both implementations consume those exact folds through their
# explicit sample-splitting APIs.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(DoubleML)
  library(mlr3)
  library(mlr3learners)
  library(data.table)
})

MODULE <- "08_dml"

df <- read_csv_strict(MODULE)

# DoubleML wants a DoubleMLData object built from a data.table.
dt <- as.data.table(df)
covariates <- c("exper", "expersq", "black", "south", "smsa")

dml_data <- DoubleML::DoubleMLData$new(
  data = dt, y_col = "lwage", d_cols = "educ", x_cols = covariates
)

# Linear regression nuisance learners (closed-form, no MC noise).
ml_g <- mlr3::lrn("regr.lm")
ml_m <- mlr3::lrn("regr.lm")

set.seed(PARITY_SEED)
dml_obj <- DoubleML::DoubleMLPLR$new(
  data    = dml_data,
  ml_l    = ml_g,
  ml_m    = ml_m,
  n_folds = 5L,
  n_rep   = 1L,
  score   = "partialling out",
  dml_procedure = "dml2"
)
fold_levels <- sort(unique(dt$fold_id))
test_ids <- lapply(fold_levels, function(k) which(dt$fold_id == k))
train_ids <- lapply(fold_levels, function(k) which(dt$fold_id != k))
smpls <- list(list(train_ids = train_ids, test_ids = test_ids))
dml_obj$set_sample_splitting(smpls)
dml_obj$fit()

theta <- as.numeric(dml_obj$coef["educ"])
se    <- as.numeric(dml_obj$se["educ"])
ci    <- dml_obj$confint(joint = FALSE, level = 0.95)["educ", ]

rows <- list(
  parity_row(
    module    = MODULE,
    statistic = "theta_DML_PLR",
    estimate  = theta,
    se        = se,
    ci_lo     = unname(ci[1]),
    ci_hi     = unname(ci[2]),
    n         = nrow(df)
  )
)

write_results(MODULE, rows,
              extra = list(dml_model = "PLR",
                           n_folds = 5L,
                           ml_g = "regr.lm",
                           ml_m = "regr.lm",
                           fold_source = "user",
                           fold_column = "fold_id",
                           score = "partialling out",
                           dml_procedure = "dml2"))
